doc/cascadia/Cascading-Default-Settings.md

```diff
@@ -0,0 +1,571 @@
+like so:
+
+```json
+// To view the default settings, open the defaults.json file in this directory
+```
+
+The "Settings" button would then only open the file the user needs to edit, and
+provide them instructions on how to open the defaults file.
+
+### How does this work with the settings UI?
+
+If we only have one version of the settings models (Globals, Profiles,
+ColorShemes, Keybindings) at runtime, and the user changes one of the settings
+with the settings UI, how can we tell that settings changed?
+
+Fortunately, this should be handled cleanly by the algorithm proposed above, in
+the "Serializing User Settings" section. We'll only be serializing settings that
+have changed from the defaults, so only the actual changes they've made will be
+persisted back to the user settings file.
+
+## Capabilities
+### Security
+
+I don't think this will introduce any new security issues that weren't already
+present
+
+### Reliability
+I don't think this will introduce any new reliability concerns that weren't
+already present
+
+### Performance, Power, and Efficiency
+
+By not writing the defaults to disk, we'll theoretically marginally improve the
+load and save times for the `profiles.json` file, by simply having a smaller
+file to load.
+
+### Accessibility
+N/A
+
+## Potential Issues
+### Migrating Existing Settings
+
+I believe that existing `profiles.json` files will smoothly update to this
+model, without breaking. While in the new model, the `profiles.json` file can be
+much more sparse, users who have existing `profiles.json` files will have full
+settings in their user settings. We'll leave their files largely untouched, as
+we won't touch keys that have the same values as defaults that are currently in
+the `profiles.json` file. Fortunately though, users should be able to remove
+much of the boilerplate from their `profiles.json` files, and trim it down just
+to their modifications.
+
+### Dynamic Profile Generators Need to be Enabled
+With the current proposal, profiles that are generated by a dynamic profile
+generator _need_ that generator to be enabled for the profile to appear in the
+list of profiles. If the generator isn't enabled, then the important parts of
+the profile (name, commandline) will never be set, and the profile's settings
+from the user settings will be ignored at runtime.
+
+For generators where the generation of profiles might be a lengthy process, this
+could negatively impact startup time. Take for example, some hypothetical
+generator that needs to make web requests to generate dynamic profiles. Because
+we need the finalized settings to be able to launch the terminal, we'll be stuck
+loading until that generator is complete.
+
+However, if the user disables that generator entirely, we'll never display that
+profile to the user, even if they've done that setup before.
+
+So the trade-off with this design is that non-existent dynamic profiles will

```

Comment:
Can you parallel the individual generators so they can do their work at the same time and WaitForMultipleObjects on all of them to be done? That might help.

Reply:
or the winrt equivalent, with coroutines and futures and stuff

Reply:
That's certainly an option, though we'd have to make sure to make the vector of profiles thread safe and all that jazz. Not impossible, so I'll leave it as a suggestion.

Reply:
nah because each dynamic generator runs in complete isolation you can fan-out, get back multiple vectors, merge them later, order them in one step when the user profiles come in.

Reply:
that makes me think: should it be the dynamic generator's job to say "save the settings?" might be overly complicated for an extension author. the profile code should _know_ that a dynamic gen produced a profile it's never seen before.

Reply:
this also removes the need to send the dynamic generator _any_ info about the user's profiles!

Reply:
So I was thinking it's not the generator's responsibility, but the app's responsibility, to look at the list of generated profiles and check if each and every one already has a entry in the user settings. So the generator will just blind hand back a list of profiles, the deserializer will figure out the rest